gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactorTimeBasedJobPropCreator.java

```diff
@@ -0,0 +1,154 @@
+import org.joda.time.format.DateTimeFormatter;
+import org.joda.time.format.PeriodFormatter;
+import org.joda.time.format.PeriodFormatterBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Optional;
+import com.google.common.collect.Lists;
+
+
+/**
+ * This class creates MR job properties for time-based compaction, i.e., the input folder pattern
+ * contains a timestamp, such as PageViewEvent/hourly/2015/05/22.
+ *
+ * The folder pattern can be configured by compaction.timebased.folder.pattern, the default value
+ * being YYYY/MM/dd, which means an MR job will be launched for each qualified folder that matches
+ * [this.inputPath]&#47;*&#47;*&#47;*.
+ *
+ * To control which folders to process, use properties compaction.timebased.max.time.ago and
+ * compaction.timebased.max.time.ago. The format is ?m?d?h, e.g., 3m or 2d10h.
+ *
+ * @author ziliu
+ */
+public class MRCompactorTimeBasedJobPropCreator extends MRCompactorJobPropCreator {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MRCompactorTimeBasedJobPropCreator.class);
+
+  private final String folderPattern;
+  private final DateTimeZone timeZone;
+  private final DateTimeFormatter timeFormatter;
+
+  static class Builder extends MRCompactorJobPropCreator.Builder<MRCompactorTimeBasedJobPropCreator.Builder> {
+    @Override
+    MRCompactorTimeBasedJobPropCreator build() {
+      return new MRCompactorTimeBasedJobPropCreator(this);
+    }
+  }
+
+  MRCompactorTimeBasedJobPropCreator(Builder builder) {
+    super(builder);
+    this.folderPattern = getFolderPattern();
+    this.timeZone =
+        DateTimeZone.forID(this.state.getProp(ConfigurationKeys.COMPACTION_TIMEZONE,
+            ConfigurationKeys.DEFAULT_COMPACTION_TIMEZONE));
+    this.timeFormatter = DateTimeFormat.forPattern(this.folderPattern).withZone(this.timeZone);
+  }
+
+  @Override
+  List<State> createJobProps() throws IOException {
+    List<State> allJobProps = Lists.newArrayList();
+    if (!fs.exists(this.topicInputDir)) {
+      LOG.warn("Input folder " + this.topicInputDir + " does not exist. Skipping topic " + topic);
+      return allJobProps;
+    }
+
+    String folderStructure = getFolderStructure();
+    for (FileStatus status : this.fs.globStatus(new Path(this.topicInputDir, folderStructure))) {
+      DateTime folderTime = getFolderTime(status.getPath());
+      if (shouldProcessFolder(status.getPath(), folderTime)) {
+        Path jobOutputDir = new Path(this.topicOutputDir, folderTime.toString(this.timeFormatter));
+        Path jobTmpDir = new Path(this.topicTmpDir, folderTime.toString(this.timeFormatter));
+        Optional<State> jobProps = createJobProps(status.getPath(), jobOutputDir, jobTmpDir);
+        if (jobProps.isPresent()) {
+          allJobProps.add(jobProps.get());
+        }
+      }
+    }
+
+    return allJobProps;
+  }
+
+  private String getFolderStructure() {
+    return this.folderPattern.replaceAll("[a-zA-Z0-9]+", "*");
```

Comment:
`data/YYYY/MM/dd/more/levels` is not a valid pattern and cannot be passed to `DateTimeFormat.forPattern()`.

I'll change `folderPattern` to `folderTimePattern` to avoid confusion.